File: 1408-find-the-smallest-divisor-given-a-threshold/find-the-smallest-divisor-given-a-threshold.java
class Solution {
    public int smallestDivisor(int[] nums, int threshold) {
        int h=0;
        int l=1;
        for(int i:nums){
            h=Math.max(h,i);
        }
        while(l<=h){
            int mid=l+(h-l)/2;
            if(possible(nums,threshold,mid)){
                h=mid-1;
            }
            else{
                l=mid+1;
            }
        }
        return l;
    }
    public boolean possible(int[] nums,int thresh,int mid){
        int sum=0;
        for(int i:nums){
            sum+=(i+mid-1)/mid;
            if(sum>thresh){
                return false;
            }
        }
        return true;
    }
}

File: 1408-find-the-smallest-divisor-given-a-threshold/find-the-smallest-divisor-given-a-threshold.cpp
class Solution {
public:
    int smallestDivisor(vector<int>& nums, int threshold) {
        int l=1;
        int h=0;
        for(int x:nums){
            h=max(h,x);
        }
        int ans=h;
        while(l<=h){
            int mid=l+(h-l)/2;
            if(poss(nums,mid,threshold)){
                ans=mid;
                h=mid-1;
            }
            else{
                l=mid+1;
            }
        }
        return ans;
    }
    bool poss(vector<int>& nums,int div,int limit){
        int sum=0;
        for(int x:nums){
            sum+=(x+div-1)/div;
            if(sum>limit){
                return false;
            }
        }
        return true;
    }
};